semgrep/semgrep/dependency_aware_rule.py

```diff
@@ -0,0 +1,108 @@
+from pathlib import Path
+from typing import Dict
+from typing import Generator
+from typing import List
+from typing import Tuple
+
+from dependencyparser.models import PackageManagers
+from dependencyparser.package_restrictions import dependencies_range_match_any
+from dependencyparser.package_restrictions import find_and_parse_lockfiles
+from dependencyparser.package_restrictions import ProjectDependsOnEntry
+
+from semgrep.error import SemgrepError
+from semgrep.rule import Rule
+from semgrep.rule_match import CoreLocation
+from semgrep.rule_match import RuleMatch
+from semgrep.verbose_logging import getLogger
+
+logger = getLogger(__file__)
+
+
+PACKAGE_MANAGER_MAP = {
+    "pypi": PackageManagers.PYPI,
+    "npm": PackageManagers.NPM,
+}
+
+
+def parse_depends_on_restrictions(
+    entries: List[List[Dict[str, str]]]
+) -> Generator[ProjectDependsOnEntry, None, None]:
+    for wrapper_entry in entries:
+        for entry in wrapper_entry:
+            # schema checks should gaurantee we have these fields, but we'll code defensively
+            namespace = entry.get("namespace")
+            if namespace is None:
+                raise SemgrepError(f"project-depends-on is missing `namespace`")
+            pm = PACKAGE_MANAGER_MAP.get(namespace)
+            if pm is None:
+                raise SemgrepError(
+                    f"unknown package namespace: {namespace}, only {list(PACKAGE_MANAGER_MAP.keys())} are supported"
+                )
+            package_name = entry.get("package")
+            if package_name is None:
+                raise SemgrepError(f"project-depends-on is missing `package`")
+            semver_range = entry.get("version")
+            if semver_range is None:
+                raise SemgrepError(f"project-depends-on is missing `version`")
+            yield ProjectDependsOnEntry(
+                namespace=pm, package_name=package_name, semver_range=semver_range
+            )
+
+
+def run_dependency_aware_rule(
+    matches: List[RuleMatch],
+    rule: Rule,
+    targets: List[Path],
+) -> Tuple[List[RuleMatch], List[SemgrepError]]:
+    """
+    Run a dependency aware rule.
```

Comment:
More detail would be helpful. Maybe would be more helpful to mention takes Matches and rule and filters out the rules that don't match the depends on + what behavior is if rule doesnt have depends-on